airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py

```diff
@@ -0,0 +1,390 @@
+    """
+
+    data_field = "transaction_details"
+    primary_key = [["transaction_info", "transaction_id"]]
+    cursor_field = ["transaction_info", "transaction_initiation_date"]
+
+    start_date_max = {"hours": 36}  # this limit is found experimentally
+    records_per_request = 10000  # API limit
+
+    def path(self, **kwargs) -> str:
+        return "transactions"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+
+        decoded_response = response.json()
+        total_pages = decoded_response.get("total_pages")
+        page_number = decoded_response.get("page")
+        if page_number < total_pages:
+            return {"page": page_number + 1}
+        else:
+            return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        page_number = 1
+        if next_page_token:
+            page_number = next_page_token.get("page")
+
+        return {
+            "start_date": stream_slice["start_date"],
+            "end_date": stream_slice["end_date"],
+            "fields": "all",
+            "page_size": self.page_size,
+            "page": page_number,
+        }
+
+    def stream_slices(
+        self, sync_mode, cursor_field: List[str] = None, stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Optional[Mapping[str, any]]]:
+        """
+        Returns a list of slices for each day (by default) between the start date and end date.
+        The return value is a list of dicts {'start_date': date_string, 'end_date': date_string}.
+
+        Slice does not cover period for slice_start_date > maximum_allowed_start_date
+        """
+        period = timedelta(**self.stream_slice_period)
+
+        slice_start_date = max(self.start_date, isoparse(stream_state.get("date")) if stream_state else self.start_date)
+
+        slices = []
+        while slice_start_date <= self.maximum_allowed_start_date:
+            slices.append(
+                {"start_date": slice_start_date.isoformat(), "end_date": min(slice_start_date + period, self.end_date).isoformat()}
+            )
+            slice_start_date += period
+
+        return slices
+
+
+class Balances(PaypalTransactionStream):
+    """
+    Stream for Balances /v1/reporting/balances
+
+    API Docs: https://developer.paypal.com/docs/integration/direct/transaction-search/#check-balances
+    """
+
+    primary_key = "as_of_time"
+    cursor_field = "as_of_time"
+    data_field = None
+
+    def path(self, **kwargs) -> str:
+        return "balances"
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        return {
+            "as_of_time": stream_slice["start_date"],
```

Comment:
i'm a little unsure about the impact of this parameter. Let's say we have 3 transactions: 

* $5, january 1st
* $10, february 10
* $20, march 15

the slices might be something like: january 1, february 1, march 1

wouldn't setting `as_of_time` to those values respectively result in us outputting: 
* $5, january 1st
* $10, february 10
* $20, march 15
* $10, february 10
* $20, march 15
* $20, march 15

Reply:
API returns account balance on a specific date

Account transactions, date:
+$5,  jan 1
+$10, feb 10
+$20, mar 15


SYNC:
start_date   = jan 1
slice_period = 1 day

Balance, at date:
$5  jan1
$5  jan2
$5  jan3
$5  jan4
..  
$5  feb9
$10 feb10 <--
$10 feb11
$10 feb12
...
$10 mar14
$20 mar15 <--
$20 mar16
$20 mar17